lib/src/ioctl/g_ext_ctrls.rs

```diff
@@ -0,0 +1,418 @@
+use bitflags::bitflags;
+use enumn::N;
+use std::convert::TryFrom;
+use std::mem;
+use std::os::unix::io::AsRawFd;
+use thiserror::Error;
+
+use crate::bindings;
+
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum ExtControlKind {
+    H264DecodeMode = bindings::V4L2_CID_STATELESS_H264_DECODE_MODE,
+    H264StartCode = bindings::V4L2_CID_STATELESS_H264_START_CODE,
+    H264Sps = bindings::V4L2_CID_STATELESS_H264_SPS,
+    H264Pps = bindings::V4L2_CID_STATELESS_H264_PPS,
+    H264ScalingMatrix = bindings::V4L2_CID_STATELESS_H264_SCALING_MATRIX,
+    H264PredWeights = bindings::V4L2_CID_STATELESS_H264_PRED_WEIGHTS,
+    H264SliceParams = bindings::V4L2_CID_STATELESS_H264_SLICE_PARAMS,
+    H264DecodeParams = bindings::V4L2_CID_STATELESS_H264_DECODE_PARAMS,
+    FwhtParams = bindings::V4L2_CID_STATELESS_FWHT_PARAMS,
+}
+
+impl From<&ExtControl> for ExtControlKind {
+    fn from(ctrl: &ExtControl) -> Self {
+        match *ctrl {
+            ExtControl::H264DecodeMode => Self::H264DecodeMode,
+            ExtControl::H264StartCode => Self::H264StartCode,
+            ExtControl::H264Sps => Self::H264Sps,
+            ExtControl::H264Pps => Self::H264Pps,
+            ExtControl::H264ScalingMatrix => Self::H264ScalingMatrix,
+            ExtControl::H264PredWeights => Self::H264PredWeights,
+            ExtControl::H264SliceParams => Self::H264SliceParams,
+            ExtControl::H264DecodeParams => Self::H264DecodeParams,
+            ExtControl::FwhtParams(_) => Self::FwhtParams,
+        }
+    }
+}
+
+#[repr(u32)]
+#[derive(Debug, Copy, Clone, PartialEq, N)]
+pub enum Colorspace {
+    Default = bindings::v4l2_colorspace_V4L2_COLORSPACE_DEFAULT,
+    Smpte170M = bindings::v4l2_colorspace_V4L2_COLORSPACE_SMPTE170M,
+    Smpte240M = bindings::v4l2_colorspace_V4L2_COLORSPACE_SMPTE240M,
+    Rec709 = bindings::v4l2_colorspace_V4L2_COLORSPACE_REC709,
+    Bt878 = bindings::v4l2_colorspace_V4L2_COLORSPACE_BT878,
+    SystemM470 = bindings::v4l2_colorspace_V4L2_COLORSPACE_470_SYSTEM_M,
+    SystemBG470 = bindings::v4l2_colorspace_V4L2_COLORSPACE_470_SYSTEM_BG,
+    Jpeg = bindings::v4l2_colorspace_V4L2_COLORSPACE_JPEG,
+    Srgb = bindings::v4l2_colorspace_V4L2_COLORSPACE_SRGB,
+    OpRgb = bindings::v4l2_colorspace_V4L2_COLORSPACE_OPRGB,
+    Bt2020 = bindings::v4l2_colorspace_V4L2_COLORSPACE_BT2020,
+    Raw = bindings::v4l2_colorspace_V4L2_COLORSPACE_RAW,
+    DciP3 = bindings::v4l2_colorspace_V4L2_COLORSPACE_DCI_P3,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Error)]
+pub enum DataConversionError {
+    #[error("Cannot convert colorspace")]
+    ColorspaceErr,
+    #[error("Cannot convert xfer function")]
+    XferFuncErr,
+    #[error("Cannot convert YCbCr encoding")]
+    YCbCrEncodingErr,
+    #[error("Cannot convert quantization")]
+    QuantizationErr,
+}
+
+impl TryFrom<u32> for Colorspace {
```

Comment:
Do we need these trivial `TryFrom` implementations? Client code can just call `Foo::n` itself and deal with the result (the different error types are not particularly helping here since client code will know which type it tries to convert to).

Reply:
Will try that in the next iteration.